File: 0876-middle-of-the-linked-list/0876-middle-of-the-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode middleNode(ListNode head) {
        // 일단, ListNode 의 head 를 이용해서, next 를 통해서 size 를 파악한다.
        int sizeOfLinkedList = 0;
        ListNode cur = head;
        
        while (true) {
            // cur 의 next 계속 호출하면서 앞으로 나아가고, sizeOfLinkedList++ 를 해줌, next 가 null 인 순간 break;
            sizeOfLinkedList++;
            if (cur.next == null) {
                break;
            }
            cur = cur.next;
        }
        
        int returnIndexInLinkedList = (sizeOfLinkedList / 2) + 1;
        int nowIndex = 1;
        cur = head;
        
        while (true) {        
            if (nowIndex == returnIndexInLinkedList) {
                return cur;
            }
            
            cur = cur.next;
            nowIndex++;   
        }
    }
}

File: 0876-middle-of-the-linked-list/0876-middle-of-the-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
  ListNode* middleNode(ListNode* head) {
    ListNode* cur=head;
    int size=0;
    while(cur!=nullptr){
      size++;
      cur=cur->next;
    }
    cur=head;
    int selectedNode=size/2+1;
    size=0;
    while(cur!=nullptr){
      size++;
      if(size==selectedNode){
        return cur;
      }
      cur=cur->next;
    }
    return nullptr;
  }
};